Mathlib/Algebra/Category/Mon/FilteredColimits.lean

```diff
@@ -0,0 +1,508 @@
+      following, we will show that this has the structure of a _commutative_ additive monoid."]
+noncomputable abbrev M : MonCat.{max v u} :=
+  MonCat.FilteredColimits.colimit.{v, u} (F ⋙ forget₂ CommMonCat MonCat.{max v u})
+set_option linter.uppercaseLean3 false in
+#align CommMon.filtered_colimits.M CommMonCat.FilteredColimits.M
+set_option linter.uppercaseLean3 false in
+#align AddCommMon.filtered_colimits.M AddCommMonCat.FilteredColimits.M
+
+@[to_additive]
+noncomputable instance colimitCommMonoid : CommMonoid.{max v u} (M.{v, u} F):=
+  { (M.{v, u} F) with
+    mul_comm := fun x y => by
+      refine Quot.induction_on₂ x y ?_
+      clear x y
+      intro x y
+      let k := max' x.1 y.1
+      let f := IsFiltered.leftToMax x.1 y.1
+      let g := IsFiltered.rightToMax x.1 y.1
+      rw [colimit_mul_mk_eq.{v, u} (F ⋙ forget₂ CommMonCat MonCat) x y k f g,
+        colimit_mul_mk_eq.{v, u} (F ⋙ forget₂ CommMonCat MonCat) y x k g f]
+      dsimp
+      rw [mul_comm] }
+set_option linter.uppercaseLean3 false in
+#align CommMon.filtered_colimits.colimit_comm_monoid CommMonCat.FilteredColimits.colimitCommMonoid
+set_option linter.uppercaseLean3 false in
+#align AddCommMon.filtered_colimits.colimit_add_comm_monoid AddCommMonCat.FilteredColimits.colimitAddCommMonoid
+
+/-- The bundled commutative monoid giving the filtered colimit of a diagram. -/
+@[to_additive "The bundled additive commutative monoid giving the filtered colimit of a diagram."]
+noncomputable def colimit : CommMonCat.{max v u} :=
+  CommMonCat.of (M.{v, u} F)
+set_option linter.uppercaseLean3 false in
+#align CommMon.filtered_colimits.colimit CommMonCat.FilteredColimits.colimit
+set_option linter.uppercaseLean3 false in
+#align AddCommMon.filtered_colimits.colimit AddCommMonCat.FilteredColimits.colimit
+
+/-- The cocone over the proposed colimit commutative monoid. -/
+@[to_additive "The cocone over the proposed colimit additive commutative monoid."]
+noncomputable def colimitCocone : Cocone F where
+  pt := colimit.{v, u} F
+  ι := { (MonCat.FilteredColimits.colimitCocone.{v, u}
+    (F ⋙ forget₂ CommMonCat MonCat.{max v u})).ι with }
+set_option linter.uppercaseLean3 false in
+#align CommMon.filtered_colimits.colimit_cocone CommMonCat.FilteredColimits.colimitCocone
+set_option linter.uppercaseLean3 false in
+#align AddCommMon.filtered_colimits.colimit_cocone AddCommMonCat.FilteredColimits.colimitCocone
+
+-- Porting note : need to add `FunLike` instance manually
+@[to_additive]
+noncomputable local instance FunLike_instance (t : Cocone.{v, max v u, v} F) :
+  FunLike (F.obj j ⟶ t.pt)
+    ((F ⋙ forget CommMonCat).obj j)
+    (fun _ => t.pt) :=
+show FunLike (F.obj j →* t.pt) ((F ⋙ forget CommMonCat.{max v u}).obj j)
+  (fun _ => t.pt) by infer_instance
+
+-- Porting note : need to add `FunLike` instance manually
+@[to_additive]
+noncomputable local instance FunLike_instance' (t : Cocone.{v, max v u, v} F) :
```

Comment:
Why are these only local instances? What is going on here, why can't they be synthesised?

Reply:
I suspect Lean don't know the arrows in category of monoids are just monoid homomorphisms. Lean knows that homomorphisms are `FunLike` but can't see that arrows are FunLike unless it is expanded manually.

Reply:
I stated `FunLike` instance for arrows in `(Comm)MonCat`in `MonCat` file, so local instances no longer needed